dali/operators/reader/loader/video/video_file.cc

```diff
@@ -0,0 +1,232 @@
+    DALI_FAIL("Could not convert frame data to RGB");
+  }
+}
+
+bool VideoFileCPU::ReadRegularFrame(uint8_t *data, bool copy_to_output) {
+  while(av_read_frame(ctx_, packet_) >= 0) {
+    if (packet_->stream_index != stream_id_) {
+      continue;
+    }
+
+    if (avcodec_send_packet(codec_ctx_, packet_) < 0) {
+      DALI_FAIL("Failed to send packet to decoder");
+    }
+
+    int ret = avcodec_receive_frame(codec_ctx_, frame_);
+
+    if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
+      continue;
+    }
+
+    if (!copy_to_output) {
+      return true;
+    }
+
+    if (sws_ctx_ == nullptr) {
+      sws_ctx_ = sws_getContext(
+        Width(), 
+        Height(), 
+        codec_ctx_->pix_fmt, 
+        Width(), 
+        Height(), 
+        AV_PIX_FMT_RGB24, 
+        SWS_BILINEAR, 
+        nullptr, 
+        nullptr, 
+        nullptr);
+      DALI_ENFORCE(sws_ctx_, "Could not create sw context");
+    }
+
+    CopyToOutput(data);
+    return true;
+  }
+
+  return false;
+}
+
+void VideoFileCPU::Reset() {
+  DALI_ENFORCE(
+    av_seek_frame(ctx_, stream_id_, 0, AVSEEK_FLAG_FRAME) >= 0,
+    make_string("Could not seek to the first frame of video ", filename_));
+  avcodec_flush_buffers(codec_ctx_);
+}
+
+void VideoFileCPU::SeekFrame(int frame_id) {
+  // TODO(awolant): Optimize seeking: 
+  //  - when we seek next frame,
+  //  - when we seek frame with the same keyframe as the current frame
+  //  - for CFR, when we know pts, but don't know keyframes 
+
+  DALI_ENFORCE(
+    frame_id >= 0 && frame_id < NumFrames(),
+    make_string("Invalid seek frame id. frame_id = ", frame_id, ", num_frames = ", NumFrames()));
+
+  auto &frame_entry = index_[frame_id];
+  int keyframe_id = frame_entry.last_keyframe_id;
+  auto &keyframe_entry = index_[keyframe_id];
+
+  // Seeking while on flush frame. Need to reset flush state
+  if (flush_state_) {
+    flush_state_ = false;
+  }
+ 
+  DALI_ENFORCE(
+    av_seek_frame(ctx_, stream_id_, keyframe_entry.pts, AVSEEK_FLAG_FRAME) >= 0,
+    make_string("Failed to seek to frame ", frame_id, "with keyframe", keyframe_id, "in video ", filename_));
+
+  avcodec_flush_buffers(codec_ctx_);
+
+  for (int i = 0; i < (frame_id - keyframe_id); ++i) {
+    ReadNextFrame(nullptr, false);
+  }
+}
+
+bool VideoFileCPU::ReadFlushFrame(uint8_t *data, bool copy_to_output) {
+  if (!flush_state_) {
+    avcodec_send_packet(codec_ctx_, nullptr);
+    flush_state_ = true;
+  }
+
+  if (avcodec_receive_frame(codec_ctx_, frame_) < 0) {
+    flush_state_ = false;
+    return false;
+  }
+
+  if (copy_to_output) {
+    CopyToOutput(data);
+  }
+
+  return true;
+}
+
+void VideoFileCPU::ReadNextFrame(uint8_t *data, bool copy_to_output) {
+  if (!flush_state_) {
+      if (ReadRegularFrame(data, copy_to_output)) {
+        return;
+      }
+  }
+  if (ReadFlushFrame(data, copy_to_output)) {
+    return;
+  }
+
+  Reset();
+
+  if (!ReadRegularFrame(data, copy_to_output)) {
+    DALI_FAIL("Error while reading frame");
+  }
```

Comment:
Not exactly sure if we should wrap up. Maybe the user should call Reset explicitly before he ca read next frame.

Reply:
Same as https://github.com/NVIDIA/DALI/pull/3362#discussion_r713796510

Reply:
I think it violates the principle of least astonishment - epsecially in something called ***File.

Reply:
Done, I make the `ReadNextFrame` return boolean and it returns false, where it hits the end of the file. I made `Reset` public, so the caller can manually reset it, if needed. Adjusted the tests to test this.